### PR TITLE
Prevent deletion

### PR DESCRIPTION
**Pull request by TalShorSap**
Prevent deletion flag on instance spec

### CODE REVIEW

File: api/v1/serviceinstance_types.go [hunk 1]
```diff
@@ -56,6 +56,11 @@ type ServiceInstanceSpec struct {
 	// +kubebuilder:default={}
 	Shared *bool `json:"shared,omitempty"`
 
+	// Indicates if the instance deletion will be prevented
+	// +optional
+	// +kubebuilder:default=false
+	PreventDeletion bool `json:"prevent_deletion"`
```
Comment (kerenlahav) [inline api/v1/serviceinstance_types.go:62]: should be an annotation and not part of the spec

File: api/v1alpha1/serviceinstance_types.go [hunk 2]
```diff
@@ -52,6 +52,11 @@ type ServiceInstanceSpec struct {
 	// +kubebuilder:default={}
 	Shared *bool `json:"shared,omitempty"`
 
+	// Indicates if the instance deletion will be prevented
+	// +optional
+	// +kubebuilder:default=false
+	PreventDeletion bool `json:"prevent_deletion"`
```
Comment (kerenlahav) [inline api/v1alpha1/serviceinstance_types.go:58]: not relevant

File: api/v1/serviceinstance_types.go [hunk 3]
```diff
@@ -121,6 +126,7 @@ type ServiceInstanceStatus struct {
 // +kubebuilder:printcolumn:JSONPath=".spec.serviceOfferingName",name="Offering",type=string
 // +kubebuilder:printcolumn:JSONPath=".spec.servicePlanName",name="Plan",type=string
 // +kubebuilder:printcolumn:JSONPath=".spec.shared",name="shared",type=boolean
+// +kubebuilder:printcolumn:JSONPath=".spec.preventDeletion",name="PreventDeletion",type=boolean
```
Comment (kerenlahav) [inline api/v1/serviceinstance_types.go:129]: remove

**Review by kerenlahav — CHANGES_REQUESTED** [review]

File: api/v1/serviceinstance_validating_webhook.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	return nil, nil
+}
+
+func (si *ServiceInstance) ValidateUpdate(old runtime.Object) (warnings admission.Warnings, err error) {
+	return nil, nil
+}
+
+func (si *ServiceInstance) ValidateDelete() (warnings admission.Warnings, err error) {
+	if si.Annotations != nil {
+		if _, ok := si.Annotations[api.PreventDeletion]; ok {
```
Comment (kerenlahav) [inline api/v1/serviceinstance_validating_webhook.go:48]: what if iset
"services.cloud.sap.com/preventDeletion": "false" ?

Comment (TalShorSap) [reply api/v1/serviceinstance_validating_webhook.go]: fixed

**Review by kerenlahav — APPROVED** [review]

**Pull request merged**